### PR TITLE
Add launcher module

### PR DESCRIPTION
**Pull request by atoulme**
This adds a simple (simplistic?) launcher library to help create applications with sane bases fast.

### CODE REVIEW

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+    /**
+     * Executes with a given set of arguments.
+     * 
+     * @param logger the logger to use for normal logging.
+     * @param out the stream to print to STDOUT.
+     * @param err the stream to print to STDERR.
+     * @param args the set of arguments provided to the program.
+     */
+    void run(Logger logger, PrintStream out, PrintStream err, String[] args);
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: `handle`

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    /**
+     * Handles an exception thrown during launch.
+     *
+     * @param logger the logger to use for normal logging.
+     * @param out the stream to print to STDOUT.
+     * @param err the stream to print to STDERR.
+     * @param throwable the exception thrown.
+     */
+    void run(Logger logger, PrintStream out, PrintStream err, Throwable throwable);
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: `handle`

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+   * The functional interface for launch exception handlers.
+   */
+  @FunctionalInterface
+  public interface LaunchExceptionHandler {
+
+    /**
+     * Handles an exception thrown during launch.
+     *
+     * @param logger the logger to use for normal logging.
+     * @param out the stream to print to STDOUT.
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: "The standard output stream." or "The stream for standard output."

File: launcher/build.gradle [hunk 4]
```diff
@@ -0,0 +1,15 @@
+description = 'A launcher library handling parameters, usage instructions and logging'
+
+dependencies {
+  compile 'org.logl:logl-api:0.1.0'
+  compile 'org.logl:logl-jul:0.1.0'
```
Comment (cleishm) [inline launcher/build.gradle]: You shouldn't include a logl backend.

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+   * @param arguments the varargs arguments
+   * @return the native array of varargs arguments
+   */
+  public static String[] args(String... arguments) {
+    return arguments;
+  }
+
+  private final Map<String, LauncherHandler> handlerMap = new HashMap<>();
+
+  private LoggerProvider loggerProvider = new JULLoggerProvider();
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: Should be injected

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+   * @param args The command line arguments.
+   */
+  public void launch(String... args) {
+    if (loggerProvider == null) {
+      throw new IllegalStateException("Logger provider has not been set");
+    }
+    if (defaultHandler == null) {
+      throw new IllegalStateException("Default handler has not been set");
+    }
+    Logger logger = loggerProvider.getLogger("launch");
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: This logger isn't used anywhere in this class, other than to pass to the handle methods. The handlers should probably get a loggerProvider, so they can create their own named loggers. And then I wonder if the handlers need to be passed it during the `handle` method, or if they can be provided the loggerProvider as a dependency injection (or captured into the closure, for a handler defined as a lambda).

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    try {
+      for (String arg : args) {
+        LauncherHandler handler = handlerMap.get(arg);
+        if (handler != null) {
+          handler.handle(logger, sysout, syserr, args);
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: sysout/syserr aren't used in this class - just passed directly to handle methods. Why if funnelling them via this class better than directly injecting into the handlers, or captured into the lambda closure?

File: launcher/src/main/java/net/consensys/cava/launcher/Launcher.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  public void launch(String... args) {
+    if (loggerProvider == null) {
+      throw new IllegalStateException("Logger provider has not been set");
+    }
+    if (defaultHandler == null) {
+      throw new IllegalStateException("Default handler has not been set");
+    }
+    Logger logger = loggerProvider.getLogger("launch");
+    try {
+      for (String arg : args) {
```
Comment (cleishm) [inline launcher/src/main/java/net/consensys/cava/launcher/Launcher.java]: This should probably use an argument parser instead or just matching strings?

**Comment by atoulme** [comment]
I think I need to rethink the need for a launcher module a bit more. I'll close this PR.

**Pull request closed without merging**